### PR TITLE
[+] Color logger

### PR DESCRIPTION
**Pull request by hykilpikonna**
* Add the basic framework for colored logging in both Minecraft and Console environments
* Add an `uncheckedMapOf()` method to create infinitely-long hash maps whereas the checked `Map.of` has limited number of arguments
* Make `ArgParser` work better with custom loggers

image

### CODE REVIEW

File: src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java:33]: I like the newline formatting. I was gonna comment on how you squished everything to one line on the lower builders but I see you've changed existing stuff too 😦 Not a big deal and definitely nothing like a merge blocker, but I'm sad.

File: src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java:71]: This can't be done inside the action because... it will continue parsing arguments? Oh I see, when we set consume arguments to false, we had `--help` get consumed as the text argument. 😳 Gotcha, nice solution!

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/utils/ColorLogger.java:76]: Does Minecraft have escape strings? What if I want to log plaintext `this&a pizza`?

A template library is usually best to format these things. Obviously with our limited time we can merge this but if this was production this would find a way to bite us pretty quickly I think.

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: Also do we have a logging library that can handle this for us? Logging libraries should be responsible to handle escape sequences in a proper way. Some terminals (notably Xcode) don't support these escapes and the result is a bunch of garbled text.

**Review by 1whatleytay — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: Check out https://github.com/HyDevelop/HyLogger (I made it lol)

Also [HyFetch can try to detect](https://github.com/hykilpikonna/hyfetch/blob/b86f43e9abc6e7b618eedb06d889fe404d55b886/hyfetch/termenv.py#L18) if a terminal supports Xterm256 or Xterm truecolor but not if it doesn't support ANSI.

But tbh there's really no way of correctly identifying if a terminal supports ANSI that works 100% of the time. There are so many different terminals out there and not all of them follow the same standard.

Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: I'll just assume that whoever is using the tool on the command line is using it with a terminal that supports colors.

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:76]: Yea in minecraft you can use `&&a` to escape `&a` but I don't have time to implement that now 😭

File: src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/client/arguments/ArgsParser.java:33]: oops sorry

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: > There are so many different terminals out there and not all of them follow the same standard.

This is why we move the responsibility to a library maintained by other people. If a terminal is not supported, then it's their fault to support it. As developers we can't possibly think of all the different environments our app will run in, but the library developers can do that thinking for us.

Again not a blocker. It's the last day before feedback lol

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:76]: Spooky.

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: Quick fix? 🤔 

```java
var ampEscape = "[gFk38N-EsCaPeD AmPeRcEnT-f47Svw]";
in = in.replace("&&", ampEscape);

// Replace colors

in = in.replace(ampEscape, "&");
```

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:76]: Quick fix? 🤔 

```java
var ampEscape = "[gFk38N EsCaPeD AmPeRsAnD f47Svw]";
in = in.replace("&&", ampEscape);

// Replace colors

in = in.replace(ampEscape, "&");
```

Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:76]: Hahaha, you don't need to... lol

Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:76]: Why do I keep commenting right after things get merged...

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (MstrPikachu) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: > If a terminal is not supported, then it's their fault to support it. 

when you use intuition to skip steps when showing work and lose marks:
"it's not my fault you need more steps"

File: src/main/java/org/hydev/mcpm/utils/Sugar.java
Comment (MstrPikachu) [inline src/main/java/org/hydev/mcpm/utils/Sugar.java:82]: Do we have to suppress any warnings here?

File: src/main/java/org/hydev/mcpm/utils/ColorLogger.java
Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/utils/ColorLogger.java:41]: Man, this is a library. It's not like omitting steps 3 through 5, it's like citing a proven theorem.

File: src/main/java/org/hydev/mcpm/utils/Sugar.java
Comment (hykilpikonna) [reply src/main/java/org/hydev/mcpm/utils/Sugar.java:82]: it's an unchecked cast. Due to type erasure, we cannot check if the keys or values are objects of type K or V, we can only put it in the precondition of the function and assume it's true. But it's still way better than Java's `Map.of` that only supports up to 10 pairs because their key and values are hardcoded

**Review by MstrPikachu — APPROVED** [review]
Good if there are no warnings

**Pull request merged**
